wrapper/src/main/java/software/amazon/jdbc/plugin/cache/DataRemoteCachePlugin.java

```diff
@@ -223,45 +231,105 @@ public <T, E extends Exception> T execute(
+
     // Query result can be served from the cache if it has a configured TTL value, and it is
     // not executed in a transaction as a transaction typically need to return consistent results.
     if (!isInTransaction && (configuredQueryTtl != null)) {
-      incrCounter(totalCallsCounter);
-      result = fetchResultSetFromCache(mainQuery);
-      if (result == null) {
-        // Cache miss. Need to fetch result from the database
-        needToCache = true;
-        incrCounter(missCounter);
-        LOGGER.finest("Got a cache miss for SQL: " + sql);
-      } else {
-        LOGGER.finest("Got a cache hit for SQL: " + sql);
-        // Cache hit. Return the cached result
-        incrCounter(hitCounter);
-        try {
-          result.beforeFirst();
-        } catch (final SQLException ex) {
-          throw WrapperUtils.wrapExceptionIfNeeded(exceptionClass, ex);
+      cacheContext = telemetryFactory.openTelemetryContext(
+          TELEMETRY_CACHE_HIT, TelemetryTraceLevel.TOP_LEVEL);
+      cacheMissContext = telemetryFactory.openTelemetryContext(
+          TELEMETRY_CACHE_MISS, TelemetryTraceLevel.TOP_LEVEL);
+      try{
+        result = fetchResultSetFromCache(mainQuery);
+        if (result == null) {
+          // Cache miss. Need to fetch result from the database
+          needToCache = true;
+          incrCounter(jdbcCacheMissCounter);
+          LOGGER.finest("Got a cache miss for SQL: " + sql);
+          if (cacheContext != null) {
+            cacheContext.setSuccess(false); // Cache miss = failure
+          }
+        } else {
+          LOGGER.finest("Got a cache hit for SQL: " + sql);
+          // Cache hit. Return the cached result
+          incrCounter(jdbcCachedQueryCounter);
+          if (cacheMissContext != null) {
+            cacheMissContext.setSuccess(false);
+            cacheMissContext.closeContext();
+            cacheMissContext = null;
+          }
+
+          try {
+            result.beforeFirst();
+            if (cacheContext != null) {
+              cacheContext.setSuccess(true); // Cache hit = success
+            }
+            return resultClass.cast(result);
+          } catch (final SQLException ex) {
+            if (cacheContext != null) {
+              cacheContext.setSuccess(false);
+              cacheContext.setException(ex);
+            }
+            throw WrapperUtils.wrapExceptionIfNeeded(exceptionClass, ex);
+          }
+        }
+      } finally {
+        if (cacheContext != null) {
+          cacheContext.closeContext();
         }
-        return resultClass.cast(result);
       }
+
+    } else {
+      incrCounter(jdbcCacheBypassCount);
     }
 
-    result = (ResultSet) jdbcMethodFunc.call();
+    dbContext = telemetryFactory.openTelemetryContext(
+        TELEMETRY_DATABASE_QUERY, TelemetryTraceLevel.TOP_LEVEL);
+    try {
+      result = (ResultSet) jdbcMethodFunc.call();
+      if (dbContext != null) {
+        dbContext.setSuccess(true);
```

Comment:
No need to indicate `dbContext` as success or failure. Just treat all DB queries the same for now.